src/Compilers/Core/Portable/Operations/BasicBlock.cs

```diff
@@ -1,152 +1,119 @@
+            {
+                _statementsBuilder.AddRange(statements);
+            }
 
-        internal void AddPredecessor(BasicBlock block)
-        {
-            _predecessors.Add(block);
-        }
+            internal void RemoveStatements()
+            {
+                _statementsBuilder.Clear();
+            }
 
-        internal void RemovePredecessor(BasicBlock block)
-        {
-            _predecessors.Remove(block);
-        }
+            internal void AddPredecessor(BasicBlock block)
+            {
+                _predecessorsBuilder.Add(block);
+            }
 
-        public enum BranchKind
-        {
-            None,
-            Regular,
-            Return,
-            StructuredExceptionHandling,
-            ProgramTermination,
-            Throw,
-            ReThrow,
-            Error,
-        }
+            internal void RemovePredecessor(BasicBlock block)
+            {
+                _predecessorsBuilder.Remove(block);
+            }
 
-        public struct Branch
-        {
-            private ImmutableArray<ControlFlowGraph.Region> _lazyFinallyRegions;
+            public struct Branch
+            {
+                private ImmutableArray<ControlFlowRegion> _lazyFinallyRegions;
 
-            public BranchKind Kind { get; internal set; }
-            public BasicBlock Destination { get; internal set; }
+                public ControlFlowBranchSemantics Kind { get; set; }
+                public BasicBlock Destination { get; set; }
 
-            /// <summary>
-            /// What regions are exited (from inner most to outer most) if this branch is taken.
-            /// </summary>
-            public ImmutableArray<ControlFlowGraph.Region> LeavingRegions { get; internal set; }
+                /// <summary>
+                /// What regions are exited (from inner most to outer most) if this branch is taken.
+                /// </summary>
+                public ImmutableArray<ControlFlowRegion> LeavingRegions { get; set; }
 
-            /// <summary>
-            /// What regions are entered (from outer most to inner most) if this branch is taken.
-            /// </summary>
-            public ImmutableArray<ControlFlowGraph.Region> EnteringRegions { get; internal set; }
+                /// <summary>
+                /// What regions are entered (from outer most to inner most) if this branch is taken.
+                /// </summary>
+                public ImmutableArray<ControlFlowRegion> EnteringRegions { get; set; }
 
-            /// <summary>
-            /// The finally regions the control goes through if the branch is taken
-            /// </summary>
-            public ImmutableArray<ControlFlowGraph.Region> FinallyRegions
-            {
-                get
+                /// <summary>
+                /// The finally regions the control goes through if the branch is taken
+                /// </summary>
+                public ImmutableArray<ControlFlowRegion> FinallyRegions
```

Comment:
>public ImmutableArray<ControlFlowRegion> FinallyRegions [](start = 16, length = 55)

I think this property and the corresponding field should be moved to ControlFlowBranch. I'll take care of that too.